sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala

```diff
@@ -312,3 +315,90 @@ case class Logarithm(left: Expression, right: Expression)
     """
   }
 }
+
+case class Round(child: Expression, scale: Expression) extends Expression {
+
+  def this(child: Expression) = {
+    this(child, Literal(0))
+  }
+
+  def children: Seq[Expression] = Seq(child, scale)
+
+  def nullable: Boolean = true
+
+  private lazy val scaleV = scale.asInstanceOf[Literal].value
+  private lazy val _scale = if (scaleV != null) scaleV.asInstanceOf[Int] else 0
+
+  override lazy val dataType: DataType = {
+    child.dataType match {
+      case StringType | BinaryType => DoubleType
+      case DecimalType.Fixed(p, s) => DecimalType(p, _scale)
+      case t => t
+    }
+  }
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    child.dataType match {
+      case _: NumericType | NullType | BinaryType | StringType => // satisfy requirement
+      case dt =>
+        return TypeCheckFailure(s"Only numeric, string or binary data types" +
+          s" are allowed for ROUND function, got $dt")
+    }
+    scale match {
+      case Literal(value, LongType) =>
+        if (value.asInstanceOf[Long] < Int.MinValue || value.asInstanceOf[Long] > Int.MaxValue) {
+          return TypeCheckFailure("ROUND scale argument out of allowed range")
+        }
+      case Literal(_, _: IntegralType) | Literal(_, NullType) => // satisfy requirement
+      case child =>
+        if (child.find { case _: AttributeReference => true; case _ => false } != None) {
+          return TypeCheckFailure("Only Integral Literal or Null Literal " +
+            s"are allowed for ROUND scale arguments, got ${child.dataType}")
+        }
+    }
+    TypeCheckSuccess
+  }
+
+  def eval(input: InternalRow): Any = {
+    val evalE = child.eval(input)
+
+    if (evalE == null || scaleV == null) return null
+
+    children(0).dataType match {
+      case decimalType: DecimalType =>
+        val decimal = evalE.asInstanceOf[Decimal]
+        if (decimal.changePrecision(decimal.precision, _scale)) decimal else null
+      case ByteType =>
+        round(evalE.asInstanceOf[Byte], _scale)
+      case ShortType =>
+        round(evalE.asInstanceOf[Short], _scale)
+      case IntegerType =>
+        round(evalE.asInstanceOf[Int], _scale)
+      case LongType =>
+        round(evalE.asInstanceOf[Long], _scale)
+      case FloatType =>
+        round(evalE.asInstanceOf[Float], _scale)
+      case DoubleType =>
+        round(evalE.asInstanceOf[Double], _scale)
+      case StringType =>
+        round(evalE.asInstanceOf[UTF8String].toString, _scale)
+      case BinaryType =>
+        round(UTF8String.fromBytes(evalE.asInstanceOf[Array[Byte]]).toString, _scale)
+    }
+  }
+
+  private def round[T](input: T, scale: Int)(implicit bdc: BigDecimalConverter[T]): T = {
+    input match {
```

Comment:
As `T` is a fix value during the runtime, make the `round` return a function instead? Probably will be more efficient in performance.